File: 2331-evaluate-boolean-binary-tree/2331-evaluate-boolean-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def evaluateTree(self, root: Optional[TreeNode]) -> bool:
        mp, stack, prev, original = {}, [], None, root
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                root = stack[-1]
                if root.right and root.right != prev:
                    root = root.right
                else:
                    if not root.left and not root.right:
                        mp[root] = root.val
                    elif root.val == 2:
                        mp[root] = mp[root.left] or mp[root.right]
                    else:
                        mp[root] = mp[root.left] and mp[root.right]
                    stack.pop()
                    prev = root
                    root = None
        return mp[original]

File: 2331-evaluate-boolean-binary-tree/2331-evaluate-boolean-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var evaluateTree = function(root) {
    if (!root) return true
    if (!root.left && !root.right) return root.val === 0 ? false : true
    if (root.val === 2) {
        return evaluateTree(root.left) || evaluateTree(root.right)
    } else {
        return evaluateTree(root.left) && evaluateTree(root.right)
    }
};

evaluateTree = function(root) {
    if (!root.left || !root.right) return root.val === 0 ? false : true
    const original = root
    const stack = []
    const map = {}
    let prev = null
    while (root || stack.length) {
        if (root) {
            stack.push(root)
            root = root.left
        } else {
            root = stack.at(-1)
            if (root.right && root.right !== prev) {
                root = root.right
            } else {
                if (!root.left && !root.right) {
                    map[root] = root.val
                } else if (root.val === 2) {
                    map[root] = map[root.left] || map[root.right]
                } else {
                    map[root] = map[root.left] && map[root.right]
                }
                stack.pop()
                prev = root
                root = null
            }
        }
    }
    return map[original]
}

evaluateTree = function(root) {
    switch (root.val) {
        case 1:
            return true
        case 2:
            return evaluateTree(root.left) || evaluateTree(root.right)
        case 3:
            return evaluateTree(root.left) && evaluateTree(root.right)
        default:
            return false
    }
}